src/shell/input/service.c

```diff
@@ -0,0 +1,198 @@
+        goto error;
+    if (!streq (ranks, "all")) {
+        /* Ensure that targeted tasks are still open.
+         * ("all" is treated as "all open")
+         */
+        if (!(ids = idset_decode (ranks)))
+            goto error;
+        if (!is_subset (in->open_tasks, ids)) {
+            errno = EPIPE;
+            goto error;
+        }
+    }
+    if (input_eventlog_put (in->shell, o) < 0)
+        goto error;
+    if (eof && subtract_idset (in->open_tasks, ranks, ids) < 0)
+        shell_log_errno ("failed to remove '%s' from open tasks", ranks);
+    if (flux_respond (h, msg, NULL) < 0)
+        shell_log_errno ("flux_respond");
+    idset_destroy (ids);
+    return;
+error:
+    if (flux_respond_error (h, msg, errno, NULL) < 0)
+        shell_log_errno ("flux_respond");
+    idset_destroy (ids);
+}
+
+struct input_service *input_service_create (flux_shell_t *shell)
+{
+    struct input_service *in;
+
+    if (!(in = calloc (1, sizeof (*in))))
+        return NULL;
+    in->shell = shell;
+    if (!(in->open_tasks = idset_create (0, IDSET_FLAG_AUTOGROW))
+        || idset_range_set (in->open_tasks,
+                            0,
+                            shell->info->total_ntasks - 1))
+        goto error;
+    if (flux_shell_service_register (in->shell,
+                                     "stdin",
+                                     input_service_stdin_cb,
+                                     in) < 0)
+        shell_die_errno (1, "flux_shell_service_register");
+
+    /* Do not add a completion reference for the stdin service, we
+     * don't care if the user ever sends stdin */
+
+    if (input_eventlog_init (shell) < 0)
+        goto error;
+
+    return in;
+error:
+    input_service_destroy (in);
+    return NULL;
+}
+
+static int input_service_init (flux_plugin_t *p,
+                               const char *topic,
+                               flux_plugin_arg_t *args,
+                               void *data)
+{
+    struct input_service *in;
+    const char *type = "service";
+    flux_shell_t *shell = flux_plugin_get_shell (p);
+
+    /* Only active on shell rank 0
+     */
+    if (shell->info->shell_rank != 0)
+        return 0;
+
+    if (flux_shell_getopt_unpack (shell,
+                                  "input",
+                                  "{s?{s?s}}",
+                                  "stdin",
+                                   "type", &type) < 0)
+        return -1;
+
+    /* Check validity of input.stdin.type here. Only valid types currently
+     * are "service" and "file":
+     */
+    if (!streq (type, "service") && !streq (type, "file"))
+        return shell_log_errn (0, "input.stdin.type=%s invalid", type);
+
+    if (!streq (type, "service"))
+        return 0;
+
+    if (!(in = input_service_create (shell)))
+        return -1;
+    if (flux_plugin_aux_set (p, "builtin.input-service", in,
+                            (flux_free_f) input_service_destroy) < 0) {
```

Comment:
Style: long parameter lists should be broken one per line

Reply:
Thanks, I tried to fix those as I went along but missed one!